### PR TITLE
open SQLite connection with SQLITE_OPEN_URI

### PR DESCRIPTION
**Pull request by euclio**
This allows the user to set options such as in-memory caching on the database.

See the documentation here for notes on backwards-compatibility: https://www.sqlite.org/uri.html.

### CODE REVIEW

**Comment by sgrif** [comment]
This looks like it has more to do with how you compiled SQLite than Diesel. Am I missing something?

**Comment by euclio** [comment]
SQLite has support for opening databases with URIs such as `file::memory:?cache=shared`, which allows you to turn various knobs while opening the database (caching, read-only, etc.). However, this support is disabled by default. An application may request URI support by supplying the correct flags to `sqlite_open_v2`. Observe that if you use that URI with diesel a file named `file::memory:?cache=shared` will be created instead of an in-memory database (at least, it will for a brewed sqlite on Mac OS).

**Comment by euclio** [comment]
Perhaps an example will make the issue clearer?

This program creates the database in memory, which is expected when using a URI.

```rust
extern crate rusqlite;

use rusqlite::Connection;

fn main() {
    Connection::open("file::memory:?cache=shared").unwrap();
}
```

On the other hand, this program creates a the database in a file called `file::memory:?cache=shared`.

```rust
extern crate diesel;

use diesel::prelude::*;

fn main() {
    SqliteConnection::establish("file::memory:?cache=shared").unwrap();
}
```

**Comment by sgrif** [comment]
Sorry for not replying to this issue sooner. I didn't mean to ignore it, but we will likely not be able to address it until after 1.0.0

File: diesel/src/sqlite/connection/raw.rs [hunk 1]
```diff
@@ -17,8 +19,11 @@ impl RawConnection {
     pub fn establish(database_url: &str) -> ConnectionResult<Self> {
         let mut conn_pointer = ptr::null_mut();
         let database_url = try!(CString::new(database_url));
-        let connection_status =
-            unsafe { ffi::sqlite3_open(database_url.as_ptr(), &mut conn_pointer) };
+        let flags =
+            SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_NOMUTEX | SQLITE_OPEN_URI;
```
Comment (sgrif) [inline diesel/src/sqlite/connection/raw.rs]: Why are we specifying `OPEN_NOMUTEX` here? It is not a flag set by `sqlite3_open`, and it seems unrelated to your change.

**Comment by sgrif** [comment]
So if I'm reading this correctly, the only time this has a change in behavior is when the path starts with `file:`? (which means this is technically a breaking change for anyone who has a file starting with `file:` but I think we can safely ignore that)

File: diesel/src/sqlite/connection/raw.rs (hunk 1 above)
Comment (euclio) [reply diesel/src/sqlite/connection/raw.rs]: My understanding is that this flag is required to use the shared memory cache mode when opening a file with URI `file::memory:?cache=shared`.

**Comment by euclio** [comment]
That's correct.

File: diesel/src/sqlite/connection/raw.rs (hunk 1 above)
Comment (sgrif) [reply diesel/src/sqlite/connection/raw.rs]: I'd prefer we hold off on making that change for the time being. We certainly can't just do it by default, but we can possibly expose an API for this in the future.

**Comment by sgrif** [comment]
@euclio Are you still interested in finishing this?

**Comment by euclio** [comment]
Yes, will address your comments within the week.

**Comment by euclio** [comment]
Addressed comments.

**Comment by vemoo** [comment]
What's the status of this PR?
I'm interested because I want to be able to create a readonly connection, and I think this is the only way to achieve it.

**Comment by sgrif** [comment]
The status appears to be that it has a failing build and needs a rebase. The test suite would need to pass before this can be further reviewed

**Comment by sgrif** [comment]
I'd also like to see a more meaningful test -- The test provided passes without this change

**Comment by euclio** [comment]
Cool, will address feedback within the week. I haven't had much impetus to work on this since my own project that uses Diesel has stalled, but if there's more interest than just me I'm happy to work on it :)

**Comment by vemoo** [comment]
Awesome, thank you!

**Comment by euclio** [comment]
@sgrif This is ready for another review. Failure on nightly is unrelated.

**Comment by sgrif** [comment]
I agree that the clippy failure seems unrelated, but it only occurs on this PR. I can't reproduce on master or locally

**Comment by sgrif** [comment]
Never mind I see now it is affecting other PRs

**Comment by joelgallant** [comment]
@euclio did you want to rebase and finish this? I could also try to finish it off

**Comment by euclio** [comment]
Thanks for the ping, @joelgallant. Rebased.

**Comment by euclio** [comment]
Ping @sgrif

**Comment by kevincox** [comment]
Any updates on this? The docs say that this is supported so the lack of support for URIs is very confusing. https://docs.rs/diesel/1.4.4/diesel/sqlite/struct.SqliteConnection.html

**Comment by weiznich** [comment]
@kevincox I would happily accept a rebased version of this PR, as long as the CI passes Otherwise a fix for the documentation would also be welcome.

**Comment by euclio** [comment]
@weiznich I went ahead and rebased this PR.

File: CHANGELOG.md [hunk 2]
```diff
@@ -39,6 +39,13 @@ for Rust libraries in [RFC #1105](https://github.com/rust-lang/rfcs/blob/master/
   functionality of `NonAggregate`. See [the upgrade
   notes](#2-0-0-upgrade-non-aggregate) for details.
 
+### Fixed
```
Comment (weiznich) [inline CHANGELOG.md:42]: There is already a `Fixed` section in line 94. Could you move this point there?
Additionally I'm not that happy with the formulation "This allows
  applications to use in-memory databases" as it is already possible to do this using `":memory:"` as connection string passed to `SqliteConnection::establish`. Maybe just drop that point and only say that we support full sqlite URIs now?

File: diesel/src/sqlite/connection/raw.rs [hunk 3]
```diff
@@ -24,8 +24,10 @@ impl RawConnection {
     pub fn establish(database_url: &str) -> ConnectionResult<Self> {
         let mut conn_pointer = ptr::null_mut();
         let database_url = CString::new(database_url.trim_start_matches("sqlite://"))?;
```
Comment (weiznich) [inline diesel/src/sqlite/connection/raw.rs:26]: Just a random idea here: Should we treat `sqlite://URI` the same as `file://URI`? If so we should change the code in such a way that we replace `sqlite://` at the beginning of a connection URI with `file://` instead of just removing that part.

Comment (euclio) [reply diesel/src/sqlite/connection/raw.rs]: Makes sense to me. Done.

File: CHANGELOG.md (hunk 2 above)
Comment (euclio) [reply CHANGELOG.md]: Done.

**Comment by euclio** [comment]
Looks like the CI failures are spurious.

**Comment by euclio** [comment]
Happy this finally got merged! Thanks :)
